File: packages/services/Car/service/src/com/android/car/hal/HidlHalPropConfig.java
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.car.hal;

import android.hardware.automotive.vehicle.V2_0.VehicleAreaConfig;
import android.hardware.automotive.vehicle.V2_0.VehiclePropConfig;

import com.android.car.CarServiceUtils;

import java.util.ArrayList;

/**
 * HidlHalPropConfig is a HalPropConfig with an HIDL backend.
 */
public final class HidlHalPropConfig extends HalPropConfig {
    private final VehiclePropConfig mConfig;

    public HidlHalPropConfig(VehiclePropConfig config) {
        mConfig = config;

        // Do some validity checks to make sure we do not return null for get functions.
        if (mConfig.areaConfigs == null) {
            mConfig.areaConfigs = new ArrayList<VehicleAreaConfig>();
        }
        if (mConfig.configString == null) {
            mConfig.configString = new String();
        }
    }

    /**
     * Get the property ID.
     */
    @Override
    public int getPropId() {
        return mConfig.prop;
    }

    /**
     * Get the access mode.
     */
    @Override
    public int getAccess() {
        return mConfig.access;
    }

    /**
     * Get the change mode.
     */
    @Override
    public int getChangeMode() {
        return mConfig.changeMode;
    }

    /**
     * Get the area configs.
     */
    @Override
    public HalAreaConfig[] getAreaConfigs() {
        int size = mConfig.areaConfigs.size();
        HalAreaConfig[] areaConfigs = new HalAreaConfig[size];
        for (int i = 0; i < size; i++) {
            areaConfigs[i] = new HidlHalAreaConfig(mConfig.areaConfigs.get(i));
        }
        return areaConfigs;
    }

    /**
     * Get the config array.
     */
    @Override
    public int[] getConfigArray() {
        return CarServiceUtils.toIntArray(mConfig.configArray);
    }

    /**
     * Get the config string.
     */
    @Override
    public String getConfigString() {
        return mConfig.configString;
    }

    /**
     * Get the min sample rate.
     */
    @Override
    public float getMinSampleRate() {
        return mConfig.minSampleRate;
    }

    /**
     * Get the max sample rate.
     */
    @Override
    public float getMaxSampleRate() {
        return mConfig.maxSampleRate;
    }

    /**
     * Converts to AIDL or HIDL VehiclePropConfig.
     */
    @Override
    public Object toVehiclePropConfig() {
        return mConfig;
    }

    /**
     * Get the string representation for debugging.
     */
    @Override
    public String toString() {
        return mConfig.toString();
    }
}

File: packages/services/Car/cpp/vhal/client/src/HidlHalPropConfig.cpp
/*
 * Copyright (c) 2022, The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "HidlHalPropConfig.h"

#include <VehicleUtils.h>

namespace android {
namespace frameworks {
namespace automotive {
namespace vhal {

using ::android::hardware::automotive::vehicle::V2_0::VehicleAreaConfig;
using ::android::hardware::automotive::vehicle::V2_0::VehiclePropConfig;

using ::android::hardware::automotive::vehicle::toInt;

HidlHalPropConfig::HidlHalPropConfig(VehiclePropConfig&& config) {
    mPropConfig = std::move(config);
    for (VehicleAreaConfig& areaConfig : mPropConfig.areaConfigs) {
        mAreaConfigs.emplace_back(std::move(areaConfig));
    }
}

int32_t HidlHalPropConfig::getPropId() const {
    return mPropConfig.prop;
}

int32_t HidlHalPropConfig::getAccess() const {
    return toInt(mPropConfig.access);
}

int32_t HidlHalPropConfig::getChangeMode() const {
    return toInt(mPropConfig.changeMode);
}

const IHalAreaConfig* HidlHalPropConfig::getAreaConfigs() const {
    return &(mAreaConfigs[0]);
}

size_t HidlHalPropConfig::getAreaConfigSize() const {
    return mAreaConfigs.size();
}

std::vector<int32_t> HidlHalPropConfig::getConfigArray() const {
    return mPropConfig.configArray;
}

std::string HidlHalPropConfig::getConfigString() const {
    return mPropConfig.configString;
}

float HidlHalPropConfig::getMinSampleRate() const {
    return mPropConfig.minSampleRate;
}

float HidlHalPropConfig::getMaxSampleRate() const {
    return mPropConfig.maxSampleRate;
}

HidlHalAreaConfig::HidlHalAreaConfig(VehicleAreaConfig&& areaConfig) {
    mAreaConfig = std::move(areaConfig);
}

int32_t HidlHalAreaConfig::getAreaId() const {
    return mAreaConfig.areaId;
}

int32_t HidlHalAreaConfig::getMinInt32Value() const {
    return mAreaConfig.minInt32Value;
}

int32_t HidlHalAreaConfig::getMaxInt32Value() const {
    return mAreaConfig.maxInt32Value;
}

int64_t HidlHalAreaConfig::getMinInt64Value() const {
    return mAreaConfig.minInt64Value;
}

int64_t HidlHalAreaConfig::getMaxInt64Value() const {
    return mAreaConfig.maxInt64Value;
}

float HidlHalAreaConfig::getMinFloatValue() const {
    return mAreaConfig.minFloatValue;
}

float HidlHalAreaConfig::getMaxFloatValue() const {
    return mAreaConfig.maxFloatValue;
}

}  // namespace vhal
}  // namespace automotive
}  // namespace frameworks
}  // namespace android
